### DISCUSSION TITLE
6.1.0-1020-rockchip kernel possible?

### DISCUSSION THREAD

**Discussion by nrpetonr** [CNflysky/redroid-rk3588#16] [General]
hi, would it be possible to add support for 6.1.0-1020-rockchip kernel on ubuntu 24.04?

https://github.com/Joshua-Riek/ubuntu-rockchip

**Comment by CNflysky** [comment]
Currently all `RK3588 Android SDKs` (can be downloaded from open internet) are based on `ACK 5.10` (`Linux 5.10`). There is no `RK3588 Android SDK` shipped with `6.1` released yet. Hope someday board vendors can release 6.1 SDK(might with AOSP 14).

**Reply by nrpetonr** [reply]
thx, overlooked that. tested with recommended specs on 5.1 kernel and all is good.
nevertheless,  container also works with 6.1 kernel, but i can not bring android UI with scrcpy, see log hereinafter. is that connected with the kernel or ?  thx for looking into it.

scrcpy 2.5 <https://github.com/Genymobile/scrcpy>
INFO: ADB device found:
INFO:     --> (tcpip)  emulator-5554                   device  redroid13_arm64
/usr/local/share/scrcpy/scrcpy-server: 1 file pushed, 0 skipped. 90.2 MB/s (69624 bytes in 0.001s)
[server] INFO: Device: [redroid] redroid redroid13_arm64 (Android 13)
INFO: Renderer: opengl
INFO: OpenGL version: 3.3 (Compatibility Profile) Mesa 23.0.0-devel
INFO: Trilinear filtering enabled
INFO: Texture: 1080x1920

**Reply by CNflysky** [reply]
> thx, overlooked that. tested with recommended specs on 5.1 kernel and all is good.
> nevertheless,  container also works with 6.1 kernel, but i can not bring android UI with scrcpy, see log hereinafter. is that connected with the kernel or ?  thx for looking into it.
> 
> scrcpy 2.5 <https://github.com/Genymobile/scrcpy>
> INFO: ADB device found:
> INFO:     --> (tcpip)  emulator-5554                   device  redroid13_arm64
> /usr/local/share/scrcpy/scrcpy-server: 1 file pushed, 0 skipped. 90.2 MB/s (69624 bytes in 0.001s)
> [server] INFO: Device: [redroid] redroid redroid13_arm64 (Android 13)
> INFO: Renderer: opengl
> INFO: OpenGL version: 3.3 (Compatibility Profile) Mesa 23.0.0-devel
> INFO: Trilinear filtering enabled
> INFO: Texture: 1080x1920
> 

scrcpy logs are not useful at all. you need provide kernel logs and redroid logs.

**Reply by CNflysky** [reply]
> here you are.
> 
> [dmesg.log](https://github.com/user-attachments/files/16445040/dmesg.log)
> [redroid.log](https://github.com/user-attachments/files/16445041/redroid.log)
> 

```
07-31 16:58:26.824  1108  1108 F DEBUG   : Abort message: 'Failed to set damage region on surface 0xb400fa5d8b641fe0, error=EGL_BAD_ACCESS'
```
EGL library incompatible. Should wait for release of 6.1 Android sdk ,in case of EGL lib is proprietary.

**Comment by SeeleVolleri** [comment]
I see couple render error,
And a 6.1.0 kernel.
Please, use a 5.10.110, Mali only works in that version.

**Comment by CNflysky** [comment]
New development: After add Android specific dma-buf feature in Armbian 6.1 kernel, redroid(lineage-20) can now work as expected. However, all stock kernels (I've look into Armbian, Ubuntu-Rockchip, OrangePi) do not include that feature. That means you will not be able to use redroid without a customized 6.1 kernel.
Later I'll publish my fork of Armbian kernel.

**Comment by CNflysky** [comment]
https://github.com/CNflysky/linux-rockchip
build with `make ARCH=arm64 CROSS_COMPILE=aarch64-linux-gnu- BRANCH=6.1.75 LOCALVERSION=-rockchip-rk3588 bindeb-pkg`

**Reply by i-moxi** [reply]
我在rk3588上尝试使用这个配置build并安装生成的deb包，通过uname -a确定kernel已经被更换
```
# DMABUF options
CONFIG_DMABUF_CACHE=y
# CONFIG_RK_DMABUF_DEBUG is not set
# CONFIG_DMABUF_PARTIAL is not set
# CONFIG_UDMABUF is not set
# CONFIG_DMABUF_MOVE_NOTIFY is not set
# CONFIG_DMABUF_DEBUG is not set
# CONFIG_DMABUF_SELFTESTS is not set
CONFIG_DMABUF_HEAPS=y
CONFIG_DMABUF_SYSFS_STATS=y
CONFIG_DMABUF_HEAPS_DEFERRED_FREE=y
CONFIG_DMABUF_HEAPS_PAGE_POOL=y
CONFIG_DMABUF_HEAPS_SYSTEM=y
CONFIG_DMABUF_HEAPS_CMA=y
# CONFIG_DMABUF_HEAPS_SRAM is not set
# CONFIG_DMABUF_HEAPS_ROCKCHIP is not set
# end of DMABUF options
CONFIG_RK_DMABUF_PROCFS=y
```
完整配置
[config.txt](https://github.com/user-attachments/files/17745245/config.txt)


仍然显示
```
FATAL: Android specific dma-buf heap device missing
FATAL: At least one of those mandatory kernel features are not met. You must find another kernel built with those features or build a customized kernel by yourself.
```

[linux-6.1.75-rockchip-rk3588_6.1.75-rockchip-rk3588-1_arm64.buildinfo.txt](https://github.com/user-attachments/files/17745234/linux-6.1.75-rockchip-rk3588_6.1.75-rockchip-rk3588-1_arm64.buildinfo.txt)

scrcpy output
```
[server] INFO: Device: [redroid] redroid redroid13_arm64 (Android 13)
[server] ERROR: Attempt to invoke virtual method 'java.lang.Class java.lang.Object.getClass()' on a null object reference
java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.Class java.lang.Object.getClass()' on a null object reference
        at com.genymobile.scrcpy.wrappers.DisplayManager.getDisplayInfo(DisplayManager.java:89)
        at com.genymobile.scrcpy.device.Device.<init>(Device.java:77)
        at com.genymobile.scrcpy.Server.scrcpy(Server.java:145)
        at com.genymobile.scrcpy.Server.internalMain(Server.java:290)
        at com.genymobile.scrcpy.Server.main(Server.java:245)
        at com.android.internal.os.RuntimeInit.nativeFinishInit(Native Method)
        at com.android.internal.os.RuntimeInit.main(RuntimeInit.java:355)
ERROR: Server connection failed
```

**Reply by CNflysky** [reply]
> 我在rk3588上尝试使用这个配置build并安装生成的deb包，通过uname -a确定kernel已经被更换
> 
> ```
> # DMABUF options
> CONFIG_DMABUF_CACHE=y
> # CONFIG_RK_DMABUF_DEBUG is not set
> # CONFIG_DMABUF_PARTIAL is not set
> # CONFIG_UDMABUF is not set
> # CONFIG_DMABUF_MOVE_NOTIFY is not set
> # CONFIG_DMABUF_DEBUG is not set
> # CONFIG_DMABUF_SELFTESTS is not set
> CONFIG_DMABUF_HEAPS=y
> CONFIG_DMABUF_SYSFS_STATS=y
> CONFIG_DMABUF_HEAPS_DEFERRED_FREE=y
> CONFIG_DMABUF_HEAPS_PAGE_POOL=y
> CONFIG_DMABUF_HEAPS_SYSTEM=y
> CONFIG_DMABUF_HEAPS_CMA=y
> # CONFIG_DMABUF_HEAPS_SRAM is not set
> # CONFIG_DMABUF_HEAPS_ROCKCHIP is not set
> # end of DMABUF options
> CONFIG_RK_DMABUF_PROCFS=y
> ```
> 
> 完整配置 [config.txt](https://github.com/user-attachments/files/17745245/config.txt)
> 
> 仍然显示
> 
> ```
> FATAL: Android specific dma-buf heap device missing
> FATAL: At least one of those mandatory kernel features are not met. You must find another kernel built with those features or build a customized kernel by yourself.
> ```
> 
> [linux-6.1.75-rockchip-rk3588_6.1.75-rockchip-rk3588-1_arm64.buildinfo.txt](https://github.com/user-attachments/files/17745234/linux-6.1.75-rockchip-rk3588_6.1.75-rockchip-rk3588-1_arm64.buildinfo.txt)
> 
> scrcpy output
> 
> ```
> [server] INFO: Device: [redroid] redroid redroid13_arm64 (Android 13)
> [server] ERROR: Attempt to invoke virtual method 'java.lang.Class java.lang.Object.getClass()' on a null object reference
> java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.Class java.lang.Object.getClass()' on a null object reference
>         at com.genymobile.scrcpy.wrappers.DisplayManager.getDisplayInfo(DisplayManager.java:89)
>         at com.genymobile.scrcpy.device.Device.<init>(Device.java:77)
>         at com.genymobile.scrcpy.Server.scrcpy(Server.java:145)
>         at com.genymobile.scrcpy.Server.internalMain(Server.java:290)
>         at com.genymobile.scrcpy.Server.main(Server.java:245)
>         at com.android.internal.os.RuntimeInit.nativeFinishInit(Native Method)
>         at com.android.internal.os.RuntimeInit.main(RuntimeInit.java:355)
> ERROR: Server connection failed
> ```

`ls /dev/dma_heap` ?

**Reply by i-moxi** [reply]
```
nanopi-m6:~:# ls /dev/dma_heap
cma  cma-uncached  system  system-dma32  system-uncached  system-uncached-dma32
```

**Reply by CNflysky** [reply]
> ```
> nanopi-m6:~:# ls /dev/dma_heap
> cma  cma-uncached  system  system-dma32  system-uncached  system-uncached-dma32
> ```

how about `docker exec redroid logcat -d` ?
better restart your container then capture logs since Android system will produce lots of logs so critical information might be overwriten

**Reply by i-moxi** [reply]
Here is log
[redroid.log.txt](https://github.com/user-attachments/files/17745852/redroid.log.txt)

**Reply by CNflysky** [reply]
> Here is log [redroid.log.txt](https://github.com/user-attachments/files/17745852/redroid.log.txt)

i dont see any clear error, would you like to join our qq channel?

**Comment by lanfanb** [comment]
Hi All,

I successfully get redroid working on rock5b board with 6.1.43 customized kernel (debian bookworm).
Thanks for the Android specific dma-buf patch set!

For those who interests, here are steps:
1) Get rock5b official debian bookworm image installed.
2) Enable Mali GPU driver overlay (refer to radxa docs, use rsetup).
3) Build customize kernel using offical bsp tool (which uses radxa kernel repo) and patch file attached.
4) Update kernerl and reboot.
5) All set.

I can only confirm that baidunetdisk is working. I have not tested other apps yet.

Note 1: The attached patch is cherry-picked from CNflysky's repo.
Note 2: Make sure CONFIG_PSI_DEFAULT_DISABLED is set to no or unset (already in attached patch).
Note 3: envcheck.sh complains about the Mali driver & firmware version mismatch. As stock driver & firmwaver version is higher than g18, thus you can ignore them.

[0001-support-redroid-rk3588.patch.gz](https://github.com/user-attachments/files/17879019/0001-support-redroid-rk3588.patch.gz)
